### DISCUSSION TITLE
.deleted: parameter not set

### DISCUSSION THREAD

**Discussion by phidebian** [ksh93/ksh#574] [Q&A] [answered]
I was disturbed at first with static scoping, I abused bash dynamic scoping (while forced to move to bash) and struggled to get passing args by ref (typeset -n, nameref) because of circular name reference. Now I reunite with ksh93 and enjoy the safety of static scoping, I was wondering about nameref circular ref, thinking it should be impossible to do in ksh93, still I tried this 

    TC$ typeset -n a=b
    TC$ typeset -n b=a
    ksh: typeset: b: invalid self reference
    TC$ typeset -n b=c
    TC$ typeset -n c=a
    TC$ echo $a
    ksh: .deleted: parameter not set
    TC$ unset a
    ksh: unset: warning: .deleted: is read only

Though completely useless example, let say it could be a programing error, I was kinda expecting the 1st error message, then after <code>typeset -n c=a</code> I was expecting may be a similar error but got nothing (like a success), then I wondered what the <code>echo $a</code> would do and to my grand surprise I got this <code>ksh: .deleted: parameter not set</code> and also the last confusing <code>ksh: unset: warning: .deleted: is read only</code>

What does it mean ? Is that expected?

**Comment by McDutchie** [comment]
IMHO, it's a bug.

The `.deleted` thing is `NullNode` in name.c. When a name reference refers to an entity that is (subsequently) unset, it is set to refer to that in `nv_delete()`. But I think it's a bug to do this when the entity referred to is another name reference, as this defeats self-reference loop detection.

Try this patch:
```diff
diff --git a/src/cmd/ksh93/sh/name.c b/src/cmd/ksh93/sh/name.c
index 3c78d2686..a797f68fc 100644
--- a/src/cmd/ksh93/sh/name.c
+++ b/src/cmd/ksh93/sh/name.c
@@ -1246,7 +1246,8 @@ void nv_delete(Namval_t* np, Dt_t *root, int flags)
 					free(rp->sub);
 				rp->sub = 0;
 				rp = dtdelete(Refdict,(void*)rp);
-				rp->np = &NullNode;
+				if(!(flags&NV_REF))
+					rp->np = &NullNode;
 			}
 		}
 	}
@@ -3369,7 +3370,7 @@ void nv_setref(register Namval_t *np, Dt_t *hp, int flags)
 	sh.instance = 0;
 	sh.last_root = root;
 	_nv_unset(np,0);
-	nv_delete(np,(Dt_t*)0,0);
+	nv_delete(np,(Dt_t*)0,NV_REF);
 	np->nvalue.nrp = sh_newof(0,struct Namref,1,sizeof(Dtlink_t));
 	np->nvalue.nrp->np = nq;
 	np->nvalue.nrp->root = hp;
```

**Comment by McDutchie** [comment]
I tested the ksh93-history versions with `nameref a=b b=c c=a`. Loop detection works properly until ksh 93u 2010-11-22, which is also the version that introduced `NullNode`. Bug confirmed.

**Comment by McDutchie** [comment, marked as answer]
Fixed in 72d895d8cb585589645cb87bdabd48cd539716b8 (dev), d9d9c50b62a9fc043eda01f6e589f4e50f2342d4 (1.0)

**Comment by phidebian** [comment]
@McDutchie  Whew, well done :) 
Cheers
